src/hypothesis/strategies.py

```diff
@@ -857,18 +857,24 @@ def recursive(base, extend, max_leaves=100):
 def permutations(values):
     """Return a strategy which returns permutations of the collection
     "values"."""
+    from hypothesis.internal.conjecture.utils import integer_range
+
     values = list(values)
     if not values:
-        return just(()).map(lambda _: [])
-
-    def build_permutation(swaps):
-        initial = list(values)
-        for i, j in swaps:
-            initial[i], initial[j] = initial[j], initial[i]
-        return initial
-    n = len(values)
-    index = integers(0, n - 1)
-    return lists(tuples(index, index), max_size=n ** 2).map(build_permutation)
+        return builds(list)
+
+    class PermutationStrategy(SearchStrategy):
+
+        def do_draw(self, data):
+            # Reversed Fisher-Yates shuffle. Reverse order so that it shrinks
+            # propertly: This way we prefer things that are lexicographically
+            # closer to the identity.
+            result = list(values)
+            for i in hrange(len(result)):
```

Comment:
Sigh. Syntastic picked an annoying time to fail me and not tell me about the missing import. Not sure why. Should be fixed now.